Recommended Tech Debt tracking enhancements (#194)
UKHO/docs: software-engineering-policies/TechnicalDebt/TechnicalDebtPolicy.md

# Technical Debt Policy

All UKHO employees engaged with digital product development should develop a basic understanding of tehcnical debt, how it's generated, it's impacts on the business and how it can be paid off.

All individuals who identify technical debt should record it as per the following strategy. Where the individual cannot, the information should be passed to the relevant team, product manager or service owner to be raised.

## Technical Debt Definition

Deficiencies in internal quality or capability that make a product/service/application more costly to modify, extend or maintain.

## Technical Debt Types

Each product **MUST** maintain a register of technical debt within it's Azure DevOps project.

> This shoudl include all types of technical debt, i.e.:

Reviewer: typo in should

* Code – Low quality code or ineffective patterns
* Knowledge – lack of documentation or inaccessible documentation
* Automation – Lack of automated tasks forcing manual intervention (Testing, deployments, backup/restore)
* Testing – Unknown or unrecorded test scenarios, lack of test coverage
* Maintenance – Out-of-support products, usually leading to security vulnerabilities
* Security - Known and exploitable vulnerabilities
* Process - enifficient or wasteful process steps, this could be related to practice or tooling

## Technial Debt Impact

Technical debt impacts various aspects of an application's life cycle, this is often dictated by the type of technical debt but common impacts you can expect from outstanding technical debt are:

* Increased time to deliver
* Unplanned work
* Inaccurate planning
* Disengaged Development teams
* Longer times to recover
* Instability

---

## Raising Technical Debt

Technical debt identified in the current sprint but not being resolved as part of present work **MUST** be raised as a PBI in the relevant area of Azure DevOps. The relevant location is based on whether the technical debt is related to a product or a team.

> Ideally even if technical debt is going to be resolved immediatly raising it is good for understanding its impact on UKHO work and identifying that technical debt was discovered.

Reviewer: typo in immediately

  * Tech Debt Impact:
    * Availability
    * Portability
    * Compatibility
    * Performance
    * Maintainability
    * Extensibility  
  * Out-Of-Support**
  * Investigate To Remove** - This tag has been used to mark a Technical Debt item that may already have been completed or is going to be irrelevant due to other work.
  * UKHO TD priority tag:
    * TD1 - High business/technical value, high risk debt that needs to be paid off ASAP (a Security tag should automatically be considered for a TD1 prioritisation).
    * TD2 - High business/technical value (cost reduction, blocker removing, maintainability improvement), lower risk, a change that should be worked on when time/opportunity permits and is not something that can be accepted or supported long term.
    * TD3 - Tech debt that has been accepted as a risk but through paying off would add value through improving usability, maintainability, reliability or performance.
    * TD4 - Accepted risk from the business, safe to leave until service reaches end of life. Worth tracking in case developers are working in the area and can complete as quick wins.

** - Denotes where the tag is appropriate

### Tech Debt Metric Strategy

> Teams **MUST** adopt the use of the TD metrics section in their PBI template. This enables the business to gather much more useful and accurate information about technical debt and empowers the automations available for tracking and monitoring. For support in adding this section to your PBI template, either request DDC to implement the change to your templates or discuss with your delivery manager.

Reviewer: Although this sentence says teams **must** use TD metrics, the screenshot later on marks it as optional. This is confusing.

Author: Would love an alternative if you have one, it is a case of MUST use it when recording TD but only useful when recording technical debt. So happy to have recommendations

Reviewer 2: Could have a separate type of work item, a custom one?

Reviewer 2: You could also have a blank template with what you need to fill in. People don't read policy so imbedding it in the process will make it more likely to happen.

Reviewer: How about naming the section in the PBI template
`Tech Debt Metrics (if applicable)`
or
`Tech Debt Metrics (Tech Debt only)`

Author: > Could have a separate type of work item, a custom one?

In regards to creating a "TD" work item, it is possible but may require the complete rewrite of portions of the automations in place now. for minimal work and maximum benefit I think Sara's point of a better name for the PBI section would be better.

Author: Updated the Image with Sara's recommendation

#### TD Principle

> (Integer) Technical Debt Principle is the estimated cost in work-hours to pay off the technical debt at the moment of identificaiton. The discovering team/individual is probably best placed to assess this though it could be put to a team decision and refined. If this is part of a legacy product and no known SME is available then this should be considered in the estimate. Don't be afraid of large values but attempt to be as accurate as possible.

Reviewer: Is work hours the same as elapsed hours? Because man hours will vary from team to team depending upon whether that team engages in solo, paired or mobbed ways of working for that particular type of work

Author: This is another difficult point. I can define in more detail what I mean by man-hours but we need a standard approach to TD tracking as it needs to be universal for the business.

Author: I think depending on how the team works they should be able to arrive at a value that is equivalent, my recommendation is put in a work-hours estimate that assumes no pairing as a starting point as there is no guarantee that the current team will be the one who eventually handles the TD.

Reviewer: I agree that going by actual time over man hours is the most consistent way to do this - would need to make this clear in the policy and the template though

# Technical Debt Policy

> All UKHO employees engaged with digital product development should develop a basic understanding of tehcnical debt, how it's generated, it's impacts on the business and how it can be paid off.

Reviewer 3: ```suggestion
All UKHO employees engaged with digital product development should develop a basic understanding of technical debt, how it's generated, its impacts on the business and how it can be paid off.
```

All individuals who identify technical debt should record it as per the following strategy. Where the individual cannot, the information should be passed to the relevant team, product manager or service owner to be raised.

## Objective

To provide the UKHO with a detailed and expansive data set about technical debt across the organisation, enabling greater business intelligence at all levels of the organisation for more informed and effective decision making.

## Technical Debt Definition

Deficiencies in internal quality or capability that make a product/service/application more costly to modify, extend or maintain.

## Technical Debt Types

> Each product **MUST** maintain a register of technical debt within it's Azure DevOps project.

Reviewer 3: ```suggestion
Each product **MUST** maintain a register of technical debt within its Azure DevOps project.
```

This should include all types of technical debt, i.e.:

* Architectural – Tightly coupled systems (lots of criss-crossed dependencies), restrictive to extension or automation
* Code – Low quality code or ineffective patterns
* Knowledge – lack of documentation or inaccessible documentation
* Automation – Lack of automated tasks forcing manual intervention (Testing, deployments, backup/restore)
* Testing – Unknown or unrecorded test scenarios, lack of test coverage
* Maintenance – Out-of-support products, usually leading to security vulnerabilities
* Security - Known and exploitable vulnerabilities

> * Process - enifficient or wasteful process steps, this could be related to practice or tooling

Reviewer 3: ```suggestion
* Process - inefficient or wasteful process steps, this could be related to practice or tooling
```

## Technial Debt Impact

Technical debt impacts various aspects of an application's life cycle, this is often dictated by the type of technical debt but common impacts you can expect from outstanding technical debt are:

* Increased time to deliver
* Unplanned work
* Inaccurate planning
* Disengaged Development teams
* Longer times to recover
* Instability

---

## Raising Technical Debt

Technical debt identified in the current sprint but not being resolved as part of present work **MUST** be raised as a PBI in the relevant area of Azure DevOps. The relevant location is based on whether the technical debt is related to a product or a team.

Ideally even if technical debt is going to be resolved immediately raising it is good for understanding its impact on UKHO work and identifying that technical debt was discovered.

> When raising Technical Debt, please add the following information to the Technical Debt PBI:

Reviewer 3: technical debt caps vs non-caps seems a bit inconsistent

### Title

```Technical Debt - {Concise outline of Technical Debt item}```

### Description

The description of the Technical Debt item must include the following as a minimum:

* The repository the Technical Debt lives within (a link to the repo would be good but the name would suffice)
* Description of the issue in as much detail as you can provide, including class/file names if this can be included
* A proposed solution or ideas for a fix **
* Cost/Benefit and risk of the Technical Debt item, simply as a "why should this be fixed?/How long would it take?"

### Tagging Strategy

* The Technical Debt PBI **MUST** be tagged with the `````Technical Debt````` tag and should have any further appropriate tags, such as the following:
  * Product Name (could be multiple, see the product list in AppRegister; [AppRegister](http://appregister/))
  * Tech Debt Type:
    * Security
    * Knowledge
    * Automation
    * Testing
    * Architecture
    * Code
    * Infrastructure
  * Tech Debt Impact:
    * Availability
    * Portability
    * Compatibility

>     * Performance

Reviewer 4: I think this still needs some sort of ility to differentiate it from the Type, so maybe this is a Type rather than an Impact and you could have Usability for the impact?

Author: The problem with TD of a performance type is that improving the performance of a solution isn't necessarily TD, it is considered just standard CI work. Now some TD could impact the performance of a solution but the performance of a solution isn't necessarily TD. Does that make sense?

Reviewer 4: So is it just "Ability"... to Perform :)

Author: Could be phrased that way but didn't think that was clear enough.

    * Maintainability
    * Extensibility  
  * Out-Of-Support**
  * Investigate To Remove** - This tag has been used to mark a Technical Debt item that may already have been completed or is going to be irrelevant due to other work.

>   * UKHO TD priority tag:

Reviewer 4: One issue we may face is object persistance, if something classified as a TD1 you have to fix it ASAP, if you arent looking at it right now, does it move to a TD2? Maybe it is a 1, but as soon as you start the project up this has to be fixed in the first sprint.

Author: So the TD tags aren't strict in that sense, it is a tag for the technically focused SMEs to use to inform others when refining or reviewing the piece of TD. This shouldn't change, as the original identifier has flagged it as TD1 so should be being looked at for resolution at every opportunity. 

It has been raised that people might just put everything as TD1 then, my counter point being, are we not meant to all be professional in our work practice?

    * TD1 - High business/technical value, high risk debt that needs to be paid off ASAP (a Security tag should automatically be considered for a TD1 prioritisation).
    * TD2 - High business/technical value (cost reduction, blocker removing, maintainability improvement), lower risk, a change that should be worked on when time/opportunity permits and is not something that can be accepted or supported long term.
    * TD3 - Tech debt that has been accepted as a risk but through paying off would add value through improving usability, maintainability, reliability or performance.
    * TD4 - Accepted risk from the business, safe to leave until service reaches end of life. Worth tracking in case developers are working in the area and can complete as quick wins.

** - Denotes where the tag is appropriate

> ### Tech Debt Metric Strategy

Reviewer 4: I think this section is a bit too much for right now, getting things Tagged with a level, type and effect I think will be  more value.

Could the Tech Debt Metric Strategy section be taken out of this PR and brought in separately with some more engagement with the teams.

Author: This PR is to provide the information for Principles to consider going forward, If you want to build a PR using this content as the next iteration of the policy then I would begin a new branch and take content across. 

I would say that in my research, I found that businesses who did not include these metric points made very little progress in their accuracy of TD tracking and thus did not make effective gains in it's management. I would champion this section as the most valuable part of this PR. While dev teams and individual engineers might not see the value here, the value to wider business planning and oversight is much greater.

Reviewer 4: That is fine, I think the other bits are value we need now, and then this bit is in the future but not to be forgotten about future, maybe we can setup some workshops so people can get used to the options as it becomes a section in the policy document.

Author: The other bits are already in place in the existing policy so that's all good. Happy to run workshops on the other sections if the principles deem it as our chosen approach as I have been told to hand this off to them.

Teams **MUST** adopt the use of the TD metrics section in their PBI template. This enables the business to gather much more useful and accurate information about technical debt and empowers the automations available for tracking and monitoring. For support in adding this section to your PBI template, either request DDC to implement the change to your templates or discuss with your delivery manager.

> #### TD Principle

Reviewer 3: ```suggestion
#### TD Principal
```

> (Integer) Technical Debt Principle is the estimated cost in work-hours to pay off the technical debt at the moment of identificaiton. The discovering team/individual is probably best placed to assess this though it could be put to a team decision and refined. If this is part of a legacy product and no known SME is available then this should be considered in the estimate. Don't be afraid of large values but attempt to be as accurate as possible.

Reviewer 3: ```suggestion
(Integer) Technical Debt Principal is the estimated cost in work-hours to pay off the technical debt at the moment of identification. The discovering team/individual is probably best placed to assess this though it could be put to a team decision and refined. If this is part of a legacy product and no known SME is available then this should be considered in the estimate. Don't be afraid of large values but attempt to be as accurate as possible.
```

#### TD Interest

> (Integer) Technical Debt Interest is the possible cost in work-hours impacted whenever the technical debt is encountered. For security technical debt this may be difficult to quantify.

Reviewer: For security, out-of-support and other "it's not hurting now but at some point it will majorly hurt later" type debt this is difficult to quantify - what do you suggest we put in this field in those cases?

#### TD Risk

> (Float, value between 0 and 1) Technical Debt Risk is the chance that the technical debt will be encountered whenever the product is worked on. For security type technical debt this should be set as a 1, it is assumed that eventually a security flaw will be exploited so must be resolved.

Reviewer 3: Coming up with a float seems a bit precise - I can judge it to have a 0.38492 chance of impact  rather than a low, medium, high or something like that?

Author: This was recommended by several papers, the idea being that you apply a percentage (represented by a number between 0 and 1, like .5 for 50%) to help identify how often the TD will occur. So 1 would be guaranteed every time work is undertaken on the product. It is also an estimate so it's not going to be perfect. I think the security bit of this statement needs removing...

Reviewer 3: Fair enough, thanks for the explanation :)

#### (Optional) Date of Failure

(Date) When packages or software versions are going out-of-support, if they are not addressed when found this should be recorded and if a date can be found for when the version goes out-of-support this should be reported here.

> **IMPORTANT** the tagging of the technical debt PBI is used for automated metric assessments vital for visibility of the UKHOs technical debt, tagging tech debt PBIs is vital for the accuracy of this process.

Reviewer 3: ```suggestion
**IMPORTANT** the tagging of the technical debt PBI is used for automated metric assessments vital for visibility of the UKHO's technical debt. Tagging tech debt PBIs is vital for the accuracy of this process.
```

![Example Tech Debt PBI](./Example_TD_V2.PNG)

---

### How this meshes with RAID

A technical debt PBI should be created if the team is not planning on addressing the technical debt within the current sprint. We advise that teams carrying out RAID analysis and create technical debt PBIs in addition to this analysis. Consider that some technical debt items have been long lived in the past this will help support or future development teams consider the technical debt for future resolution and when planning.

## Refining Technical Debt

If the technical debt item doesn't meet any of the above it should be considered as debt with little or no value to resolution and considered for removal from the backlog.

If during refinement an imminent issue is discovered (e.g. old Technical Debt, with high risk, which is time sensitive, but has not been addressed) this should be refined and flagged to a senior Solution Architect, Product Manager and relevant Delivery Manager.

When the technical debt PBI has been refined (i.e. each item on this page has been considered and applied) ensure that you add the tag ```TD-APPROVED``` to show this.

## Technical Debt Tracking/Monitoring

> There is automated near-realtime monitoring of technical debt available for all projects within the Ukhydro Azure DevOps tenant. This is made available by the DDC technical debt prometheus exporter and relys on accurately tracked technical debt PBIs. For information on this exporter and the technology used please refer to the [technical debt monitoring](../TechnicalDebt/TechnicalDebtMonitoring.md) document.

Reviewer 3: ```suggestion
There is automated near-realtime monitoring of technical debt available for all projects within the UKHydro Azure DevOps tenant. This is made available by the DDC technical debt Prometheus exporter and relies on accurately tracked technical debt PBIs. For information on this exporter and the technology used please refer to the [technical debt monitoring](../TechnicalDebt/TechnicalDebtMonitoring.md) document.
```

>
>Learn one of the ways [GOV.UK managed technical debt](https://insidegovuk.blog.gov.uk/2013/12/10/paying-down-technical-debt-in-the-departments-and-policy-publishing-platform/).

## Additional Reading

​<https://martinfowler.com/bliki/TechnicalDebt.html>
<https://martinfowler.com/bliki/EstimatedInterest.html>
<https://martinfowler.com/bliki/TechnicalDebtQuadrant.html>

## Legacy Technical Debt Tracking

There **MUST BE** a centralised shared query in each project area that can be run in Azure Dev Ops (TFS) and Azure Dev Ops Server (VSTS) which centralises all tagged technical debt PBIs into one result set. This pulls from all accessible backlogs.

Customer:

[Link to internal queries - UKHO access only](https://github.com/UKHO/docs-internal/blob/main/links.md)

### Team Technical Debt queries

Each team should have its own query. This should be named ```Technical Debt - Basic``` and be stored in a ```Shared Queries\Technical Debt``` folder. This will ensure consistency across all teams.

## Points of Contact

> * Policy owner  - Matthew Slade

Reviewer 4: ```suggestion
* Policy Owner - Principal Support Developer
```
